cpp/src/arrow/engine/logical_plan.h

```diff
@@ -0,0 +1,136 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "arrow/type_fwd.h"
+#include "arrow/util/compare.h"
+#include "arrow/util/variant.h"
+
+namespace arrow {
+
+namespace dataset {
+class Dataset;
+}
+
+namespace engine {
+
+class Catalog;
+class Expr;
+class ExprType;
+
+class LogicalPlan : public util::EqualityComparable<LogicalPlan> {
+ public:
+  explicit LogicalPlan(std::shared_ptr<Expr> root);
+
+  const std::shared_ptr<Expr>& root() const { return root_; }
+  const ExprType& type() const;
+
+  bool Equals(const LogicalPlan& other) const;
+  std::string ToString() const;
+
+ private:
+  std::shared_ptr<Expr> root_;
+};
+
+struct LogicalPlanBuilderOptions {
+  /// Catalog containing named tables.
+  std::shared_ptr<Catalog> catalog;
+};
+
+class LogicalPlanBuilder {
```

Comment:
Hmm... I suppose `LogicalPlanBuilder` will grow a `Build` method to build a `LogicalPlan`? How will it work? The builder doesn't seem to keep track of any state.